Add support for Byte Literals

vexx32 opened the pull request:
## PR Summary

* Adds `y` suffix that is used to specify a numeric literal as the `sbyte` data type.
  * Can be combined with the existing `u` suffix as `uy` to specify the `byte` data type
* Adds tests to ensure the parsing for these works correctly.

Examples:
```powershell
PS> 100y
100
PS> 100y.GetType().FullName
System.SByte
PS> 100uy
100
100uy.GetType().FullName
System.Byte
PS> 300uy
# Error: Invalid numeric literal
```

N.B.: I believe this will be a breaking change, as it means some command names (specifically, numerals suffixed with a y such as `100y`) will now be interpreted as numerals instead of potential function or command names.

Ref: #7557 -- This will resolve the first part of that issue. The second portion of that issue will be solved in a follow-up PR where binary literal support is added.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/2965
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
Add NumberSuffixFlag enty for SignedByte

vexx32 committed:
Add explicit names for flag combinations

vexx32 committed:
Add CharTraits for typesuffix on `y`

vexx32 committed:
Add ScanNumberHelper cases for byte suffixes
Use specific combined flag names in parsing cases
Add (s)byte parsing cases for TryGetNumberValue

vexx32 committed:
Add unsigned byte tests

vexx32 committed:
Add signed byte tests

vexx32 committed:
Fix spacing issue

iSazonov commented:
@vexx32 Please open Issue in docs repo and add link to the PR description

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3355,6 +3365,15 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
+                        case NumberSuffixFlags.SignedByte:
+                            if (sbyte.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out sbyte sb))
+                            {
+                                result = (sbyte)(sb * multiplier);
```
Seems we could return false if multiplier > 1 and SignedByte or UnsignedByte. I do not know whether it makes sense to do this test explicitly. Certainly this will eliminate the exception.

vexx32 replied:
Well, still needs to be parsed as (s)byte regardless, no?

But yes, taking the shortcut of skipping that if there is a multiplier applied here is a code call, thank you!

iSazonov replied:
>  still needs to be parsed as (s)byte regardless

I think yes. It allow "recovery" and continue parsing.

vexx32 replied:
Aaaand patched in, thank you!

vexx32 replied:
And again, good catch.

The multiplier can be *negative* in valid signed parsing cases because the hex parsing ignores negative sign, deferring that to the multiplier to be applied after parsing of the hex literal. I assume this is because the hex parsing methods of `.TryParse()` do not allow negative signs, which is probably because hex parsing permits entering signed values by utilising the sign bit.

iSazonov replied:
It is not obvious behavior so please add the comment in code.

iSazonov replied:
Iexpect `result = sb;`

vexx32 replied:
Impossible to completely ignore multiplier; if a hex literal is negative, multiplier will be `-1` by this point thanks to code referenced above. 

We must still either multiply by multiplier value or else account for it in a more complex expression, e.g., `result = multiplier < 0 ? -sb : sb`

I think simply multiplying by the already-verified multiplier value is the most straightforward method.

iSazonov replied:
I agrre with this in the PR but I'd consider the possible optimization in follow PR.

vexx32 replied:
Yeah, this will be resolved more cleanly in the follow up PR. 😄

iSazonov commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -709,6 +709,32 @@ foo``u{2195}abc
+                @{ Script = "0b0y"; ExpectedValue = "0"; ExpectedType = [sbyte] }
+                @{ Script = "0b10y"; ExpectedValue = "2"; ExpectedType = [sbyte] }
+                @{ Script = "-0b10y"; ExpectedValue = "-2"; ExpectedType = [sbyte] }
+                @{ Script = "0b11111111y"; ExpectedValue = "-1"; ExpectedType = [sbyte] }
```
Please remove the tests for binary suffix.

vexx32 replied:
Done! 😄

vexx32 committed:
Remove inapplicable binary parsing tests

vexx32 committed:
Shortcut parsing fail for (s)byte when multiplier is set

Remove impossible tests
Rework byte parsing shortcut

Fix failing test scenario caused by hex parse handling of multiplier

Simplify return by ignoring multuplier for (s)byte-suffixed values

Add comment for negative multiplier possibility

Still need to utilise multiplier for signed byte case
Multiplier can determine negative in signed byte cases with hex

Adjust byte parse shortcuts
Add comments

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3355,6 +3365,15 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
+                            result = null;
+                            return false;
+                        case NumberSuffixFlags.SignedByte:
+                            if (Math.Abs(multiplier) == 1L && sbyte.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out sbyte sb))
```
Below we have such code - please add the comment there too.

Why we use `1L`? I think we should use "universal" const `1` without suffix. Below too.

vexx32 replied:
Different comment there, but yes.

Wasn't sure if c# would care about type of numeric value when testing for equality, actually. :)

vexx32 commented:
CodeFactor issues remaining are just in missing documentation in the enum; I can either add that to this PR or address it in follow-up PR with the rest of the changes. I would prefer to keep this one minimal, however. 😄

iSazonov commented:
@vexx32 Feel free to address CodeFactor issues.

vexx32 committed:
Address CodeFactor issues
Add summary headers to enum and values

vexx32 commented:
Rebased for CIs && added `<summary>` headers for CodeFactor

iSazonov commented:
@daxian-dbw Could you please review the next step PR? After that we could continue and implement previously planned performance optimizations for the code in follow PR.

vexx32 commented:
@TravisEz13 you fellas are free to pore over it one more time, but just in case you weren't aware / didn't recall, it has been discussed a good bit and reviewed by the committee in terms of the concept over in #7557 😄

vexx32 commented:
@TravisEz13  sorry for the confusion; I requested the committee further review the ensuing discussion regarding possibly supporting underscore literal syntax and native BigInteger values. This portion was reviewed, aye. 😃

iSazonov commented:
@daxian-dbw Could you please look the PR before we merge?
